### ISSUE TITLE
Jenkins plugins are not updated in existing pods

### ISSUE THREAD

**Issue by stefanlack** [opendevstack/ods-core#411]
**Describe the bug**
When a new jenkins master image is build with updated plugin versions and the image is deployed to the project, the new plugin versions are not used. Instead, the old plugin versions are loaded

**To Reproduce**
1. Update plugins.txt to use new versions
2. Build jenkins master in global cd project
3. Wait until trigger redeploys jenkins master in a project specific cd - namespace
4. Login to jenkins and verify, that plugin has expected version

**Expected behavior**
Version correct.

**Affected version (please complete the following information):**
 - OpenShift:  3.11
 - OpenDevStack 1.2, 2

**Additional context**
To solve the issue, one have to set the environment variable `OVERRIDE_PV_PLUGINS_WITH_IMAGE_PLUGINS=true`

See https://github.com/openshift/jenkins#environment-variables for details

Also `OVERRIDE_PV_CONFIG_WITH_IMAGE_CONFIG=true` may be needed.

One problem with this approach is, that jenkins looses history after restart.

Does you also examined this problem and how to you work around the problem?

**Comment by michaelsauter** [comment]
I think we face the same issues. Not sure how to tackle this right now.

**Comment by gerardcl** [comment]
Question: do we really need a PVC? shall we just make Jenkins stateless? If restarted it will anyway recover the jobs required from openshift pipelines.
What is the real need to having `/var/lib/jenkins` persisted?

**Comment by michaelsauter** [comment]
Hmm, good question. I think build numbers would be affected at least?

**Comment by gerardcl** [comment]
that is true, the counter is restarted
@clemensutschig do you see an issue on this? or any other?

**Comment by michaelsauter** [comment]
Note that this ONLY affects plugins in `plugins.txt`. Plugins shipped in the RedHat-distributed image do get updated.

**Comment by michaelsauter** [comment]
Setting `OVERRIDE_PV_PLUGINS_WITH_IMAGE_PLUGINS` allows to just delete / overwrite plugins without loosing the whole history: https://github.com/openshift/jenkins/blob/ae52cc141e027903576e71f3c038260a3599b82f/2/contrib/s2i/run#L499-L505

**Comment by clemensutschig** [comment]
@michaelsauter I remember you fixed this at bi x.. adding to ods 3?

**Comment by michaelsauter** [comment]
@clemensutschig Still in the process. The whole situation is a bit complex:

* RedHat and CentOS based images differ in how they handle plugins
* RedHat based images ship with some plugins. A symlink to those plugins is established, meaning the plugins will get updated when RedHat ships a new base image. UNLESS someone manually updated the plugins, then the link is broken and the plugin will not be updated. Unfortunately, that "someone" can also be us - when we install a plugin via the `plugins.txt`, that plugin and any of its dependencies break the symlink and therefore won't be updated.
* I do not know yet how CentOS works in detail, but it does not seem to use symlinks

@oalyman and I concluded so far:
* Remove as many plugins as possible from our `plugins.txt` - most are not needed actually as the base image comes with them already
* For RedHat, it probably makes sense to force the symlinks for all shipped plugins, overriding what was manually installed

That doesn't solve that plugins only defined in `plugins.txt` do not get updated yet. Maybe we would need to pull something off like RedHat - installing the plugins in the image, and then symlinking ...

**Comment by clemensutschig** [comment]
I took a look thru the codebase - and couldn#t we call install_plugins to force the update? .... https://github.com/openshift/jenkins/blob/ae52cc141e027903576e71f3c038260a3599b82f/2/contrib/s2i/run#L133

**Comment by michaelsauter** [comment]
@clemensutschig yes something like this will work. @oalyman and I have looked through this quite a bit today. Unfortunately it is not straight-forward to understand why things are done the way they are. Further, there are significant differences between CentOS and RHEL based images (e.g. it seems that CentOS does not use symlinks, RHEL installs plugins via RPM, ...). We just want to make sure that we do not kill any plugins installed by the user manually. We did the some fixes in #672, and now we can tackle the issue of updating "our" plugins in a new PR.
